app.js

```diff
@@ -504,27 +464,58 @@ io.sockets.on('connection', function (socket) {
                         logger.error('openHAB-cloud: Error saving connect event: ' + error);
                     }
                 });
-                notifyOpenHABStatusChange(openhab, 'online');
-            } else {
-                openhab.openhabVersion = socket.handshake.openhabVersion;
-                openhab.clientVersion = socket.handshake.clientVersion;
-                openhab.save(function (error) {
-                    if (error) {
-                        logger.error('openHAB-cloud: Error saving openhab: ' + error);
-                    }
-                });
-            }
-            socket.openhabUuid = openhab.uuid;
-            socket.openhabId = openhab.id;
+                
+                //notify user that connection is online
+                if(prevStatus === 'offline' && Date.now() - prevLastOnline > offlineNotificationTime){
+                    notifyOpenHABStatusChange(openhab, 'online');
+                } else if(prevStatus === 'online') {
+                    logger.warn('openHAB-cloud: connected openhab ' + socket.handshake.uuid + ' was previously marked as online')
+                }
+            });
         } else {
             if (error) {
                 logger.error('openHAB-cloud: Error looking up openHAB: ' + error);
             } else {
                 logger.warn('openHAB-cloud: Unable to find openHAB ' + socket.handshake.uuid);
             }
+            socket.disconnect();
         }
     });
 
+    socket.on('disconnect', function () {
+        logger.info('openHAB-cloud: Disconnected uuid ' + socket.handshake.uuid + ' connectionId ' + socket.connectionId);
+        Openhab.setOffline(socket.connectionId, function (error, openhab) {
+            if (error) {
+                logger.error('openHAB-cloud: Error saving openHAB disconnect: ' + error);
+                return;
+            }
+            if(openhab) {
+                //we will try and notifiy users of being offline
+                offlineOpenhabs[openhab.uuid] =  {
+                    date: Date.now(),
+                    connectionId: socket.connectionId
+                }
+
+                var disconnectevent = new Event({
+                    openhab: openhab.id,
+                    source: 'openhab',
+                    status: 'offline',
+                    color: 'bad'
+                });
+                
+                disconnectevent.save(function (error) {
+                    if (error) {
+                        logger.error('openHAB-cloud: Error saving disconnect event: ' + error);
+                    }
+                });
+
+                //actually would redis be better to store?  How would we coordinate who send notification?
```

Comment:
Indeed is this notification sending a completely independent task? It could be triggered by
Regularly, checking redis for instances to notify?

Reply:
So, that was a note i left to my self when running through the code, did not mean to check it in :-)

If i was starting from scratch or doing a major refactor, then yes i would be using  redis (or some queue/messaging like backend)  to persist this kind of stuff in a more distributed friendly fashion.  I again restrained from rewriting too much to keep the changes small so if something does go wrong it's easier to debug, and quicker to get this out.

Reply:
Agree, makes sense not to bundle it in here.